tools/generate_tables_sql_and_testNames.php

```diff
@@ -1,118 +1,234 @@
+                fwrite(
+                    $fp,
+                    "INSERT INTO instrument_subtests "
+                    ."(Test_name, Subtest_name, Description)"
+                    ." VALUES ('$sql_query_table_name', '"
+                    .$sql_query_table_name."_page".($page_number+1)
+                    ."', '$page');\n"
+                );
+            }
+        }
+        // Writes the error message to the sql file.
+        if (!empty($error_message)) {
+            fwrite(
+                $fp,
+                "/* The SQL file is invalid! \n"
+                ."tools/generate_tables_sql_and_testNames.php error_message:\n"
+                ."$error_message\n"
+                ."*/"
+            );
+        }
+        fclose($fp);
     }
-    $output.="PRIMARY KEY  (`CommentID`)\n) ENGINE=InnoDB DEFAULT CHARSET=utf8;\n";
-    print "Filename: $filename\n";
-    $fp=fopen($filename, "w");
-    fwrite($fp, $output);
-    fwrite($fp, "REPLACE INTO test_names (Test_name, Full_name, Sub_group) VALUES ('$tablename', '$title', 1);\n");
-    foreach($pages as $pageNo => $page) {
-        fwrite($fp, "INSERT INTO instrument_subtests (Test_name, Subtest_name, Description) VALUES ('$tablename', '" . $tablename  . "_page" . ($pageNo+1) . "', '$page');\n");
-
+    // Execute the SQL tasks if no error_message and CLI flag set.
+    if ($db->isConnected() 
+        && empty($error_message)
+        && $execute_mysql_query
+    ) {
+        if (!$db->tableExists($sql_query_table_name)) {
+            // Table doesn't exist!
+            $sql_query_statement .= implode($sql_query_columns);
+            $sql_query_statement .=
+                "PRIMARY KEY  (`CommentID`)\n) ENGINE=InnoDB "
+                ."DEFAULT CHARSET=utf8;\n";
+            $statement            = $db->prepare($sql_query_statement);
+            $statement->execute();
+        } else {
+            $error_message = "There is already an instrument installed with 
+                this name.";
+        }
+        $statement = $db->prepare(
+            'REPLACE INTO test_names (Test_name, Full_name, Sub_group)'
+            .' VALUES (?, ?, 1)'
+        );
+        $statement->bindParam(1, $sql_query_table_name, PDO::PARAM_STR);
+        $statement->bindParam(2, $title, PDO::PARAM_STR);
+        $statement->execute();
+        // Each page requires its own seprate query to be run.
+        foreach ($pages as $page_number => $page) {
+            $statement    = $db->prepare(
+                'INSERT INTO instrument_subtests'
+                .' (Test_name, Subtest_name, Description)'
+                .' VALUES (?, ?, ?)'
+            );
+            $subtest_name = $sql_query_table_name . "_page" . ($page_number+1);
+            $statement->bindParam(1, $sql_query_table_name, PDO::PARAM_STR);
+            $statement->bindParam(2, $subtest_name, PDO::PARAM_STR);
+            $statement->bindParam(3, $page, PDO::PARAM_STR);
+            $statement->execute();
```

Comment:
why are you not using a regular insert command ?

Reply:
@ridz1208 please see the comments above for more detail about the PR. The changes were to prevent sql injection and by using the PDO parameters with binding to a statement.

Reply:
to my knowledge `insert()` already does that

Reply:
I believe the injection vector was in the table name or column name which `insert()` doesn't handle.

See https://redmine.cbrain.mcgill.ca/issues/14390